Apple AppStore Apps Bridge

captn3m0 opened the pull request:
This provides a RSS feed of version changes of any iDevice application. Currently tested with the following apps:

iPhone: https://apps.apple.com/us/app/id310633997
Mac https://apps.apple.com/us/app/id982486948
iPad: https://apps.apple.com/us/app/procreate/id425073498

`appletv` and `web` are present in responses, but I couldn't find any applications for those, so those are not officially supported right now.

TODO:

- [x] Test with other platforms
- [x] check style guide
- [x] make sure it looks good. see if additional content stuffing is needed
~(screenshots?)~. `Just `nl2br` was sufficient.

<summary>

The title of each "item" in the feed intentionally contains the Application Name. 

<details>
I made the mistake of omitting application names while working on a different RSS project previously (https://github.com/captn3m0/opml-gen) which tracked GitHub Release feeds (quite similar content). The issue becomes when you get multiple feed items with title like the following:

- 2.3.0
- 1.2.5
- 1.2.1rc1

This becomes very problematic in most feed readers, especially if all these feeds are within the same category/tag, which is fairly common.

Instead, adding the application names makes it like:

- WhatsApp Messenger - 2.3.0
- Red Dead Redemption - 1.2.2
- Procreate Final - 6.3.1

which is much more readable.

</details>

I hope this will be useful to people:

- tracking iOS/Mac applications for parity against Android
- wanting to set up automated alerts for version updates on iDevice apps

captn3m0 committed:
[AppleAppStore] Initial work on the Apple App Store bridge

This brige allows you to follow iOS/iPad application updates
directly over RSS. This allows you to get notified of an application
update even if you don't have a iDevice. This may be useful in cases
where you want to be notified of a competitor's application for eg.

I built this after I got tired of waiting for WhatsApp to push their
security release on iOS. It shows up on the AppStore 7 days late.

WTF Facebook/Apple!

somini committed:
Tweak AppStore Bridge code

somini commented:
Nice bridge, congratulations.

To actually run this, I had to do some tweaks, available here: https://github.com/somini/rss-bridge/tree/pr/1316

The big problem I detected was transforming the platform into the JSON "platform". For example, the outer URL needs "iphone", but the inner JSON platform is "ios". I can't seem to find this information anywhere else.

captn3m0 commented:
Thanks, I'll merge in your changes soon.

captn3m0 committed:
[AppleAppStore] Fixes iOS parsing

captn3m0 committed:
[AppleAppStore] Adds support for Mac-only applications

- Also adds style fixes

logmanoriginal commented on bridges/AppleAppStoreBridge.php:
```diff
@@ -0,0 +1,143 @@
+
+	const PARAMETERS = array(array(
+		'id' => array(
+			'name' 			=> 'Application ID',
```
Please generally don't align arrays. It just looks wrong depending on editor settings.

logmanoriginal commented on bridges/AppleAppStoreBridge.php:
```diff
@@ -0,0 +1,143 @@
+	 */
+	private function getDataFromShoebox($id, $platform, $country){
+		$uri = $this->makeHtmlUrl($id, $country);
+		$html = getSimpleHTMLDOM($uri, 3600);
```
`3600` is not a valid header. Is this perhaps supposed to cache the data?

```suggestion
		$html = getSimpleHTMLDOMCached($uri, 3600);
```

logmanoriginal commented on bridges/AppleAppStoreBridge.php:
```diff
@@ -0,0 +1,143 @@
+	private function getJWTToken($id, $platform, $country){
+		$uri = $this->makeHtmlUrl($id, $country);
+
+		$html = getSimpleHTMLDOM($uri, 3600);
```
Same as above

```suggestion
		$html = getSimpleHTMLDOMCached($uri, 3600);
```

logmanoriginal commented on bridges/AppleAppStoreBridge.php:
```diff
@@ -0,0 +1,143 @@
+			"Authorization: Bearer $token",
+		];
+
+		$json = json_decode(getSimpleHTMLDOM($uri, $headers), true);
```
I don't think simplehtmldom is of any use here.

```suggestion
		$json = json_decode(getContents($uri, $headers), true);
```

logmanoriginal commented on bridges/AppleAppStoreBridge.php:
```diff
@@ -0,0 +1,143 @@
+		switch($platform) {
+			case 'mac':
+				return $data['relationships']['platforms']['data'][0]['attributes']['versionHistory'];
+				break;
```
This `break;` is unnecessary because of the previous `return`

```suggestion
```

logmanoriginal commented on bridges/AppleAppStoreBridge.php:
```diff
@@ -0,0 +1,143 @@
+			default:
+				$token = $this->getJWTToken($id, $platform, $country);
+				$data = $this->getAppData($id, $platform, $country, $token);
+				break;
```
`break` is generally not needed for `default`

```suggestion
```

logmanoriginal reviewed (changes requested):
Thanks for your contribution :+1: 

This bridge should support [`getName()`](https://github.com/RSS-Bridge/rss-bridge/wiki/The-getName-function) to have the name of the App in the title of the feed.

Find below a few points for consideration.

captn3m0 committed:
Switches to getSimpleHTMLDOMCached

Co-Authored-By: LogMANOriginal <LogMANOriginal@users.noreply.github.com>

captn3m0 committed:
[AppleAppStore] Adds support for getName

- getName relies on collectData running before getName (it works so far)
- Fixes getSimpleHTMLDOMCached usage
- Alignment fixes
- Removes a few redundant break statements
- Switches to getContents

logmanoriginal commented:
Array elements are still indented, but I'll merge anyway. It is probably best if we integrate a unit test for this anyway.
